gensim/corpora/text_processing_pool.py

```diff
@@ -0,0 +1,160 @@
+import multiprocessing as mp
+from multiprocessing import util
+from multiprocessing.pool import MaybeEncodingError, mapstar
+
+
+class TextProcessor(mp.Process):
+    """Generic process whose target function processes batches of texts."""
+
+    def __init__(self, group=None, target=None, name=None, args=(), kwargs={},
+                 state_kwargs=None):
+        super(TextProcessor, self).__init__(group, target, name, args, kwargs)
+        state_kwargs = {} if state_kwargs is None else dict(state_kwargs)
+        self.init_state(state_kwargs)
+
+    def init_state(self, state_kwargs):
+        for name, value in state_kwargs.items():
+            setattr(self, name, value)
+
+    def process(self, text):
+        """Process a document text; subclasses should override."""
+        raise NotImplementedError
+
+    def run(self):
+        inqueue, outqueue, initializer, initargs, maxtasks = self._args
+
+        assert maxtasks is None or (type(maxtasks) == int and maxtasks > 0)
+        put = outqueue.put
+        get = inqueue.get
+        if hasattr(inqueue, '_writer'):
+            inqueue._writer.close()
+            outqueue._reader.close()
+
+        if initializer is not None:
+            initializer(*initargs)
+
+        completed = 0
+        while maxtasks is None or (maxtasks and completed < maxtasks):
+            try:
+                task = get()
+            except (EOFError, OSError):
+                util.debug('worker got EOFError or OSError -- exiting')
+                break
+
+            if task is None:
+                util.debug('worker got sentinel -- exiting')
+                break
+
+            job, i, func, args, kwds = task
+            try:
+                if func is mapstar:
+                    args = [(self.process, arg_list) for _, arg_list in args]
+                else:
+                    func = self.process
+                result = (True, func(*args, **kwds))
+            except Exception as e:
+                result = (False, e)
+            try:
+                put((job, i, result))
+            except Exception as e:
+                wrapped = MaybeEncodingError(e, result[1])
+                util.debug("Possible encoding error while sending result: %s" % (
+                    wrapped))
+                put((job, i, (False, wrapped)))
+
+            task = job = result = func = args = kwds = None
+            completed += 1
+        util.debug('worker exiting after %d tasks' % completed)
+
+
+class _PatchedPool(mp.pool.Pool):
```

Comment:
Aha, thanks, that's interesting. @gojomo @menshikh-iv can you please review this extended multiprocessing logic?

I'm curious whether others do it this way too, since this seems a very common use-case.